File: src/codex_team/daemon/digest.py
"""Digest helpers for Codex notifications and thread items."""

from __future__ import annotations

import json as _json
import re
from pathlib import Path
from typing import Any

from codex_team.config import DigestCfg
from codex_team.schemas.turn_summary import DigestLine, DigestLineKind, TurnSummary, TurnTier

_FENCED = re.compile(r"```.*?```", re.DOTALL)


def _unwrap_item(item: Any) -> Any:
    return getattr(item, "root", item)


def _truncate(text: str, limit: int) -> str:
    if len(text) <= limit:
        return text
    indicator = f" … (truncated, {len(text)} chars)"
    return text[: max(limit - len(indicator), 0)] + indicator


def _first_line(text: str) -> str:
    return text.splitlines()[0] if text else ""


def _tail_lines(text: str, n: int) -> str:
    lines = text.splitlines()
    return "\n".join(lines[-n:]) if lines else ""


def digest_item(item: Any, cfg: DigestCfg) -> DigestLine | None:
    item = _unwrap_item(item)
    item_type = getattr(item, "type", None)
    if item_type == "commandExecution":
        return _digest_command(item, cfg)
    if item_type == "fileChange":
        return _digest_file_change(item)
    if item_type == "agentMessage":
        return DigestLine(kind=DigestLineKind.agent_message, text=getattr(item, "text", "") or "")
    if item_type == "reasoning":
        if not cfg.reasoning_capture:
            return None
        summary = getattr(item, "summary", []) or []
        if isinstance(summary, list):
            text = " ".join(str(part) for part in summary if part)
        else:
            text = str(summary)
        return DigestLine(kind=DigestLineKind.agent_message, text=text)
    if item_type == "webSearch":
        return DigestLine(kind=DigestLineKind.web_search, text=getattr(item, "query", "") or "")
    if item_type in {"mcpToolCall", "dynamicToolCall"}:
        return _digest_tool_call(item)
    if item_type == "collabAgentToolCall":
        tool = getattr(item, "tool", "") or "subagent"
        return DigestLine(kind=DigestLineKind.collab_agent, text=f"subagent={tool}")
    return None


def _digest_command(item: Any, cfg: DigestCfg) -> DigestLine:
    raw = getattr(item, "command", "") or ""
    shown = _first_line(raw)
    if len(raw) > cfg.command_truncate_chars or "\n" in raw:
        shown = _truncate(shown, cfg.command_truncate_chars)
    exit_code = getattr(item, "exit_code", None)
    duration_ms = getattr(item, "duration_ms", None)
    stderr = getattr(item, "stderr", "") or ""
    stderr_tail = None
    if exit_code not in (None, 0):
        tail = _tail_lines(stderr, cfg.stderr_tail_lines_on_fail)
        stderr_tail = tail or None
    return DigestLine(
        kind=DigestLineKind.command,
        text=shown,
        exit_code=exit_code,
        duration_ms=duration_ms,
        stderr_tail=stderr_tail,
    )


def _digest_file_change(item: Any) -> DigestLine:
    changes = getattr(item, "changes", []) or []
    first = changes[0] if changes else None
    path = getattr(first, "path", "") if first else ""
    lines_added = getattr(first, "lines_added", 0) if first else 0
    lines_removed = getattr(first, "lines_removed", 0) if first else 0
    return DigestLine(
        kind=DigestLineKind.file_change,
        text=f"{path} (+{lines_added}/-{lines_removed})",
        path=path,
        lines_added=lines_added,
        lines_removed=lines_removed,
    )


def _digest_tool_call(item: Any) -> DigestLine:
    server = getattr(item, "server", "") or ""
    tool = getattr(item, "tool", "") or ""
    args = getattr(item, "args", "") or ""
    args_head = str(args)[:80] if args else ""
    label = f"{server}/{tool}" if server else tool
    text = f"{label}({args_head})" if args_head else label
    return DigestLine(kind=DigestLineKind.tool_call, text=text, tool_name=label)


def _has_question(message: str) -> bool:
    if not message:
        return False
    stripped = _FENCED.sub("", message).rstrip()
    return stripped.endswith("?")


def classify_tier(lines: list[DigestLine], *, status: str, final_message: str | None) -> TurnTier:
    if status not in {"ok", "completed"}:
        return TurnTier.attn
    if any(line.kind == DigestLineKind.command and line.exit_code not in (None, 0) for line in lines):
        return TurnTier.attn
    if _has_question(final_message or ""):
        return TurnTier.attn
    if any(line.kind == DigestLineKind.file_change for line in lines):
        return TurnTier.normal
    return TurnTier.trivial


def build_turn_summary(
    *,
    session: str,
    turn_id: str,
    elapsed_ms: int,
    status: str,
    lines: list[DigestLine],
    final_message: str | None,
    usage_last: int | None = None,
    usage_total: int | None = None,
    error_message: str | None = None,
    completed_at: str | None = None,
) -> TurnSummary:
    files_added = sum(line.lines_added or 0 for line in lines if line.kind == DigestLineKind.file_change)
    files_removed = sum(line.lines_removed or 0 for line in lines if line.kind == DigestLineKind.file_change)
    tier = classify_tier(lines, status=status, final_message=final_message)
    return TurnSummary(
        session=session,
        turn_id=turn_id,
        elapsed_ms=elapsed_ms,
        status=status,
        tier=tier,
        final_message=final_message,
        files_added=files_added,
        files_removed=files_removed,
        lines=lines,
        usage_last_tokens=usage_last,
        usage_total_tokens=usage_total,
        error_message=error_message,
        completed_at=completed_at,
    )


def _format_line(line: DigestLine) -> str:
    if line.kind == DigestLineKind.command:
        exit_code = line.exit_code
        duration_ms = line.duration_ms or 0
        status = "ok" if exit_code == 0 else f"FAIL exit={exit_code}"
        suffix = f"\n    stderr: {line.stderr_tail}" if line.stderr_tail else ""
        return f"- [{status} {duration_ms}ms] {line.text}{suffix}"
    if line.kind == DigestLineKind.file_change:
        return f"- M {line.path} (+{line.lines_added}/-{line.lines_removed})"
    if line.kind == DigestLineKind.agent_message:
        return f"- msg: {line.text}"
    if line.kind == DigestLineKind.tool_call:
        return f"- tool: {line.text}"
    if line.kind == DigestLineKind.web_search:
        return f"- search: {line.text}"
    return f"- {line.text}"


def write_history_md(path: Path, summary: TurnSummary) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    parts = [
        f"\n## Turn {summary.turn_id} · {summary.elapsed_ms}ms · status={summary.status} · tier={summary.tier.value}\n"
    ]
    file_lines = [line for line in summary.lines if line.kind == DigestLineKind.file_change]
    if file_lines:
        parts.append("\n### File changes\n")
        parts.append("\n".join(_format_line(line) for line in file_lines))
        parts.append("\n")
    command_lines = [line for line in summary.lines if line.kind == DigestLineKind.command]
    if command_lines:
        parts.append("\n### Commands\n")
        parts.append("\n".join(_format_line(line) for line in command_lines))
        parts.append("\n")
    message_lines = [line for line in summary.lines if line.kind == DigestLineKind.agent_message]
    if message_lines:
        parts.append("\n### Messages\n")
        parts.append("\n".join(_format_line(line) for line in message_lines))
        parts.append("\n")
    if summary.final_message:
        quoted = summary.final_message.replace("\n", "\n> ")
        parts.append("\n### Final answer\n")
        parts.append(f"> {quoted}\n")
    with path.open("a", encoding="utf-8") as handle:
        handle.write("".join(parts))


def write_turns_jsonl(path: Path, summary: TurnSummary) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open("a", encoding="utf-8") as handle:
        handle.write(summary.model_dump_json() + "\n")

File: plugins/codex-team/src/digest.ts
import fs from "node:fs";
import path from "node:path";

import { DigestConfig } from "./config";
import { ensureDirFor, rotateFileIfNeeded } from "./fileIO";
import { DigestLine, TurnSummary, TurnTier } from "./models";

const FENCED_BLOCK = /```.*?```/gs;

function truncate(text: string, limit: number): string {
  if (text.length <= limit) {
    return text;
  }
  const indicator = ` ... (truncated, ${text.length} chars)`;
  return `${text.slice(0, Math.max(0, limit - indicator.length))}${indicator}`;
}

function firstLine(text: string): string {
  return text.split(/\r?\n/, 1)[0] || "";
}

function tailLines(text: string, count: number): string {
  const lines = text.split(/\r?\n/);
  return lines.slice(Math.max(0, lines.length - count)).join("\n");
}

export function digestItem(item: Record<string, unknown>, cfg: DigestConfig): DigestLine | null {
  const itemType = String(item.type ?? "");
  if (itemType === "commandExecution") {
    return digestCommand(item, cfg);
  }
  if (itemType === "fileChange") {
    return digestFileChange(item);
  }
  if (itemType === "agentMessage") {
    return { kind: "agent_message", text: String(item.text ?? "") };
  }
  if (itemType === "reasoning") {
    if (!cfg.reasoningCapture) {
      return null;
    }
    const summary = Array.isArray(item.summary) ? item.summary.map(String).join(" ") : String(item.summary ?? "");
    return { kind: "agent_message", text: summary };
  }
  if (itemType === "webSearch") {
    return { kind: "web_search", text: String(item.query ?? "") };
  }
  if (itemType === "mcpToolCall" || itemType === "dynamicToolCall") {
    return digestToolCall(item, cfg);
  }
  if (itemType === "collabAgentToolCall") {
    return { kind: "collab_agent", text: `subagent=${String(item.tool ?? "subagent")}` };
  }
  return null;
}

function digestCommand(item: Record<string, unknown>, cfg: DigestConfig): DigestLine {
  const raw = String(item.command ?? "");
  let shown = firstLine(raw);
  if (raw.length > cfg.commandTruncateChars || raw.includes("\n")) {
    shown = truncate(shown, cfg.commandTruncateChars);
  }
  const exitCode = item.exitCode == null ? null : Number(item.exitCode);
  const stderr = String(item.aggregatedOutput ?? "");
  return {
    kind: "command",
    text: shown,
    exitCode,
    durationMs: item.durationMs == null ? null : Number(item.durationMs),
    stderrTail: exitCode == null || exitCode === 0 ? null : tailLines(stderr, cfg.stderrTailLinesOnFail),
  };
}

function digestFileChange(item: Record<string, unknown>): DigestLine {
  const changes = Array.isArray(item.changes) ? item.changes : [];
  const first = (changes[0] as Record<string, unknown> | undefined) || {};
  const pathValue = String(first.path ?? "");
  const linesAdded = Number(first.linesAdded ?? first.lines_added ?? 0);
  const linesRemoved = Number(first.linesRemoved ?? first.lines_removed ?? 0);
  return {
    kind: "file_change",
    text: `${pathValue} (+${linesAdded}/-${linesRemoved})`,
    path: pathValue,
    linesAdded,
    linesRemoved,
  };
}

function digestToolCall(item: Record<string, unknown>, cfg: DigestConfig): DigestLine {
  const server = String(item.server ?? "");
  const tool = String(item.tool ?? "");
  const argsRaw = item.arguments ?? item.args ?? "";
  let argsText = "";
  try {
    argsText = typeof argsRaw === "string" ? argsRaw : JSON.stringify(argsRaw);
  } catch {
    argsText = String(argsRaw);
  }
  const argsHead = argsText ? truncate(argsText, cfg.toolArgsTruncateChars) : "";
  const label = server ? `${server}/${tool}` : tool;
  return {
    kind: "tool_call",
    text: argsHead ? `${label}(${argsHead})` : label,
    toolName: label,
  };
}

function hasQuestion(message: string): boolean {
  const stripped = message.replace(FENCED_BLOCK, "").trimEnd();
  return stripped.endsWith("?");
}

export function classifyTier(lines: DigestLine[], status: string, finalMessage: string | null): TurnTier {
  if (!["ok", "completed"].includes(status)) {
    return "attn";
  }
  if (lines.some((line) => line.kind === "command" && line.exitCode != null && line.exitCode !== 0)) {
    return "attn";
  }
  if (finalMessage && hasQuestion(finalMessage)) {
    return "attn";
  }
  if (lines.some((line) => line.kind === "file_change")) {
    return "normal";
  }
  return "trivial";
}

export function buildTurnSummary(input: Omit<TurnSummary, "filesAdded" | "filesRemoved" | "tier">): TurnSummary {
  const filesAdded = input.lines
    .filter((line) => line.kind === "file_change")
    .reduce((sum, line) => sum + Number(line.linesAdded || 0), 0);
  const filesRemoved = input.lines
    .filter((line) => line.kind === "file_change")
    .reduce((sum, line) => sum + Number(line.linesRemoved || 0), 0);
  return {
    ...input,
    filesAdded,
    filesRemoved,
    tier: classifyTier(input.lines, input.status, input.finalMessage),
  };
}

function formatLine(line: DigestLine): string {
  if (line.kind === "command") {
    const status = line.exitCode === 0 ? "ok" : `FAIL exit=${line.exitCode}`;
    const suffix = line.stderrTail ? `\n    stderr: ${line.stderrTail}` : "";
    return `- [${status} ${line.durationMs || 0}ms] ${line.text}${suffix}`;
  }
  if (line.kind === "file_change") {
    return `- M ${line.path || ""} (+${line.linesAdded || 0}/-${line.linesRemoved || 0})`;
  }
  if (line.kind === "agent_message") {
    return `- msg: ${line.text}`;
  }
  if (line.kind === "tool_call") {
    return `- tool: ${line.text}`;
  }
  if (line.kind === "web_search") {
    return `- search: ${line.text}`;
  }
  return `- ${line.text}`;
}

export function writeHistoryMd(filePath: string, summary: TurnSummary, cfg?: DigestConfig): void {
  ensureDirFor(filePath);
  if (cfg) {
    rotateFileIfNeeded(filePath, cfg.historyRotationMb);
  }
  const parts: string[] = [
    `\n## Turn ${summary.turnId} · ${summary.elapsedMs}ms · status=${summary.status} · tier=${summary.tier}\n`,
  ];
  const fileLines = summary.lines.filter((line) => line.kind === "file_change");
  if (fileLines.length > 0) {
    parts.push("\n### File changes\n");
    parts.push(fileLines.map(formatLine).join("\n"));
    parts.push("\n");
  }
  const commandLines = summary.lines.filter((line) => line.kind === "command");
  if (commandLines.length > 0) {
    parts.push("\n### Commands\n");
    parts.push(commandLines.map(formatLine).join("\n"));
    parts.push("\n");
  }
  const messageLines = summary.lines.filter((line) => line.kind === "agent_message");
  if (messageLines.length > 0) {
    parts.push("\n### Messages\n");
    parts.push(messageLines.map(formatLine).join("\n"));
    parts.push("\n");
  }
  if (summary.finalMessage) {
    parts.push("\n### Final answer\n");
    parts.push(`> ${summary.finalMessage.replace(/\n/g, "\n> ")}\n`);
  }
  fs.appendFileSync(filePath, parts.join(""), "utf8");
}

export function writeTurnsJsonl(filePath: string, summary: TurnSummary, cfg?: DigestConfig): void {
  ensureDirFor(filePath);
  if (cfg) {
    rotateFileIfNeeded(filePath, cfg.historyRotationMb);
  }
  fs.appendFileSync(filePath, `${JSON.stringify(summary)}\n`, "utf8");
}
